src/TransponderCode.hpp

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The XCSoar Project
+
+#pragma once
+
+#include "util/CharUtil.hxx"
+#include "util/NumberParser.hpp"
+
+#include <cassert>
+#include <cstdint>
+#include <cstdlib>
+#include <cstdio>
+
+/**
+ * This class stores a transponder code.
+ * It accepts the code "as is" and converts it to an octal (base 8) 
+ * representation of the code. The output needs to be formatted with %04o to
+ * get a usable transponder code.
+ */
+class TransponderCode {
+  uint_least16_t value;
+
+public:
+  /**
+   * Uninitialized.
+   */
+  TransponderCode() = default;
+
+  /**
+   * Construct an empty instance.  Its IsDefined() method will return
+   * false.
+   */
+  static constexpr TransponderCode Null() noexcept {
+    return TransponderCode(010000);
+  }
+
+  explicit constexpr TransponderCode(uint_least16_t code) noexcept {
+    SetCode(code);
+  }
+
+  explicit TransponderCode(const char* code) noexcept {
+    SetCode(code);
+  }
+
+  void SetCode(uint_least16_t code) noexcept {
+    char code_str[32];
+    snprintf(code_str, sizeof(code_str), "%u", code);
+    value = ParseCode(code_str);
```

Comment:
All you did here was eliminate the heap allocation and use a fixed-size stack buffer. But this is still bad. I mean, is `value=code` really too simple for your taste of code?

Reply:
> No, I would like to do that, but then `TransponderCode(7000)` would not be a valid code because

Of course it's not. But `TransponderCode(07000)` is.

Reply:
Yes, indeed. But a device or airspace file will most probably omit the leading `0`. I had a look last week on how to make 7000 a 07000 but failed (I only get back to formatting with `%05u`, which is similar to what the class does now). Can you give me a hint where to look at? Happy to dig deeper.

Reply:
> But a device or airspace file will most probably omit the leading `0`.

Of course the airspace file will not have a fifth digit, because that's the wrong syntax - you're mixing up syntaxes. The leading zero is C++ syntax for octal literals. To parse from a file, which is not C++ syntax, just parse the four digits as octal, like I suggested, without requiring a leading zero.

Requiring an extra zero digit would require writing extra code, but you have a history of making things more complicated than they should be.... relax your mind, do away with this complicated sh*t that occupies your mind, make it simple!